src/projection.jl

```diff
@@ -0,0 +1,62 @@
+using LinearAlgebra: Diagonal, diag
+
+"""
+    project([T::Type], x, dx)
+
+"project" `dx` onto type `T` such that it is the same size as `x`. If `T` is not provided,
+it is assumed to be the type of `x`.
```

Comment:
What's the point of `T` in addition to `x`? It sounds like you have something in mind but I don't see what. In what directions may they differ? May `T` be abstract?

Reply:
Oh T could be Tangent as in line 60

Reply:
But why? And when, I mean for which `x` would this be chosen somehow? Sounds a bit like an orthogonal feature.

Reply:
My initial thought was to ditch `T` altogether, and privilege some sort of tangent representation as canonical.

However, somewhere in the many threads about this, @oxinabox argued that we need to go both ways (to tangent and to natural differential say). e.g. if a rule author writes a rule only for `Diagonal` and wants to make sure that the `Tangent` is being transformed right

Reply:
OK, but this still seems a bit like it's rolling two things together. One is a projection to preserve the input's `x`'s real-ness / Diagon-ality etc. This would be applied automatically to every single rule.

The other is a way to massage certain `dx`s to a form digestible by certain rules. It's possible that there is some efficiency gained by fusing these processes into one function, but... for a first stab, it seems that combing them is adding complication?

Reply:
> This would be applied automatically to every single rule.

No, I don't think so.
As i see it:

In that use `project` is one way that let's us write things that can run on anything, while getting out the type (and associated basis) that we want.
But it's not the only way, and its not always the best way.
In fact i think it rarely is, since it often results in allocating and computing things that we are going going to throw away.
Its just that it is the most generic way.
Alternatives for example, are things like structure preserving maps and broadcasts.
Which avoid allocations and computing things they are going to throw away.
Though i guess in those cases you can still `project` after, but it will be a identity operation.
Like `convert(Array, [1,2,3,])`  is an identity operation.

Which brings me around to the general thing:
This is like `convert` except that it is for vector spaces and is aware of vector space concerns, where are your zeros (which is same as basis)
Which is the same convert-like operation one want to massage on the way in.

Reply:
> in those cases you can still project after, but it will be a identity operation

Yes, that's my thinking. In my list of examples somewhere, any operation mixing real & complex arrays is something which ought to project. But making every rule do the ideal thing would be a lot of work for little benefit. Whereas projecting back to a real array for `hcat([1,2], [3+im, 4+im])`'s gradient at least means it propagates no further.

Reply:
yeah, anyway, point is it is our vector space aware `convert` so it is the right name for both the thing you (often) want to do at the end.
But it is also the one you might want to do at the start